server/src/main/java/org/elasticsearch/search/rank/RankBuilder.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.rank;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.VersionedNamedWriteable;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.search.SearchService;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * {@code RankContextBuilder} is used as a base class to manage input, parsing,
+ * and subsequent generation of appropriate contexts for handling searches that
+ * require multiple queries for global rank relevance.
+ *
+ * This class contains a {@code List<QueryBuilder>}s, size, and from members that
+ * sub-classes use to execute queries to generate ranking results. These are not part
+ * of ranking input, but instead generated from their respective values within a search
+ * query. We cannot generate the list of queries until after any kNN queries are executed.
+ */
+public abstract class RankBuilder<RCB extends RankBuilder<RCB>> implements VersionedNamedWriteable, ToXContentObject {
```

Comment:
I found this to be common pattern in our builders in general. Seems nice for avoiding casting, but I'm happy to remove the generic type if you want.

Reply:
True, it is one of those common patterns that I find misuse generics :) How does it help in testing? I may have missed that.

Reply:
For integration tests, it's nice to set window_size and rank_constant, but I think you're right that in a comment below -- this can all effectively be part of an immutable object through a constructor. I will give that a try.

Reply:
Removed the generic type.